### PR TITLE
Port chaos lost documents fix to 3.6.

### PR DESCRIPTION
**Pull request by neunhoef**

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11769/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11811/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11812/

**Review by goedderz — APPROVED** [review]
LGTM

File: arangod/Aql/EngineInfoContainerCoordinator.cpp [hunk 1]
```diff
@@ -93,7 +94,7 @@ Result EngineInfoContainerCoordinator::EngineInfo::buildEngine(
     double ttl = query.queryOptions().ttl;
     TRI_ASSERT(ttl > 0);
     try {
-      queryRegistry->insert(_id, &query, ttl, true, false);
+      queryRegistry->insert(_id, &query, ttl, true, false, cluster::CallbackGuard());
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/Aql/EngineInfoContainerCoordinator.cpp:97]: This is only legacy code, as far as I see. It is the `/_api/aql/instantiate` API which was pre 3.5.

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11835/

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/13492/

**Review by KVS85 — CHANGES_REQUESTED** [review]
Please observe reproducible failures for MMFiles at `http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/13490/EDITION=community,STORAGE_ENGINE=mmfiles,TEST_SUITE=cluster,limit=(linux&&test)%7C%7Cgce/artifact/testfailures.txt`.

File: arangod/Aql/RestAqlHandler.cpp [hunk 2]
```diff
@@ -275,7 +309,8 @@ bool RestAqlHandler::registerSnippets(
     VPackSlice const snippetsSlice, VPackSlice const collectionSlice,
     VPackSlice const variablesSlice, std::shared_ptr<VPackBuilder> const& options,
     std::shared_ptr<transaction::Context> const& ctx, double const ttl,
-    SerializationFormat format, bool& needToLock, VPackBuilder& answerBuilder) {
+    SerializationFormat format, bool& needToLock, VPackBuilder& answerBuilder,
+    RebootId rebootId, std::string coordinatorId) {
```
Comment (jsteemann) [inline arangod/Aql/RestAqlHandler.cpp:313]: Can be passed by const reference, but this is an unimportant optimization here.
```suggestion
    RebootId rebootId, std::string const& coordinatorId) {
```

File: arangod/Aql/RestAqlHandler.h [hunk 3]
```diff
@@ -109,7 +110,8 @@ class RestAqlHandler : public RestVocbaseBaseHandler {
                         std::shared_ptr<arangodb::velocypack::Builder> const& options,
                         std::shared_ptr<transaction::Context> const& ctx,
                         double const ttl, aql::SerializationFormat format,
-                        bool& needToLock, arangodb::velocypack::Builder& answer);
+                        bool& needToLock, arangodb::velocypack::Builder& answer,
+                        RebootId rebootId, std::string coordinatorId);
```
Comment (jsteemann) [inline arangod/Aql/RestAqlHandler.h:114]: Can be passed by const reference, but this is an unimportant optimization here.
```suggestion
                        RebootId rebootId, std::string const& coordinatorId);
```

File: arangod/Cluster/Maintenance.cpp [hunk 4]
```diff
@@ -447,10 +480,10 @@ arangodb::Result arangodb::maintenance::diffPlanLocal(
     auto const& dbname = pdb.key.copyString();
     if (!local.hasKey(dbname)) {
       if (errors.databases.find(dbname) == errors.databases.end()) {
-        actions.emplace_back(
-            ActionDescription({{std::string(NAME), std::string(CREATE_DATABASE)},
-                               {std::string(DATABASE), std::string(dbname)}},
-                              HIGHER_PRIORITY));
+        actions.emplace_back(ActionDescription(
+              std::map<std::string, std::string>{{std::string(NAME), std::string(CREATE_DATABASE)}, {std::string("tick"), std::to_string(TRI_NewTickServer())},
```
Comment (jsteemann) [inline arangod/Cluster/Maintenance.cpp:484]: The "tick" value here will always be unique and will not be equal to the database's id. What's its purpose?
I think it deserves a code comment.

File: arangod/Cluster/Maintenance.cpp [hunk 5]
```diff
@@ -462,10 +495,10 @@ arangodb::Result arangodb::maintenance::diffPlanLocal(
   for (auto const& ldb : VPackObjectIterator(local)) {
     auto const& dbname = ldb.key.copyString();
     if (!plan.hasKey(std::vector<std::string>{DATABASES, dbname})) {
-      actions.emplace_back(
-          ActionDescription({{std::string(NAME), std::string(DROP_DATABASE)},
-                             {std::string(DATABASE), std::string(dbname)}},
-                            HIGHER_PRIORITY));
+      actions.emplace_back(ActionDescription(
+          std::map<std::string, std::string>{{std::string(NAME), std::string(DROP_DATABASE)}, {std::string("tick"), std::to_string(TRI_NewTickServer())},
```
Comment (jsteemann) [inline arangod/Cluster/Maintenance.cpp:499]: The "tick" value here will always be unique and will not be equal to the database's id. What's its purpose?
I think it deserves a code comment.

File: arangod/Aql/RestAqlHandler.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/RestAqlHandler.cpp]: done

File: arangod/Aql/RestAqlHandler.h (hunk 3 above)
Comment (jsteemann) [reply arangod/Aql/RestAqlHandler.h]: done

File: arangod/Cluster/Maintenance.cpp (hunk 5 above)
Comment (jsteemann) [reply arangod/Cluster/Maintenance.cpp]: tick attribute removed

File: arangod/Cluster/Maintenance.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/Cluster/Maintenance.cpp]: tick attribute removed

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11869/

**Review by jsteemann — APPROVED** [review]

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11874/

**Comment by KVS85** [comment]
Tests blue.

**Review by KVS85 — APPROVED** [review]

**Pull request merged**